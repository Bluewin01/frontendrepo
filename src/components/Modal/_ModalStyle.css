* {
  font-size: 18px;
}

.btn {
  text-transform: unset !important;
}

.fullScreenModal {
  min-width: 100%;
  min-height: 100%;
  margin: 0px;
}

.modalHeader {
  padding: 20px 10px 20px;
  display: relative;
  text-align: center;
  font-weight: 500;
}

.modal-body {
  padding: 20px;
}

.modal-content {
  border-radius: 10px;
}

.Button {
  border-radius: 10px;
  font-size: 18px;
  margin: 5px;
  min-width: 70px;
}
