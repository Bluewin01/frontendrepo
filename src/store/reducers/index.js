import { combineReducers } from "redux";
import { configureStore } from "@reduxjs/toolkit";
import lang from "../Lang/LangReducer";
import loader from "../Loader/LoaderReducer";
import Auth from "../Auth/AuthReducer";
import Template from "../Template/TemplateReducer";

const rootReducer = combineReducers({
  lang,
  loader,
  Auth,
  Template,
});

export const setupStore = (preloadedState) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export default rootReducer;
