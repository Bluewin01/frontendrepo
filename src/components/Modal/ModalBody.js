import React from "react";
import { Modal, Form } from "react-bootstrap";
import "./_ModalStyle.css";

export const ModalBody = (props) => {
  const { content } = props;
  return (
    <Modal.Body>
      <Form>{content}</Form>
    </Modal.Body>
  );
};
