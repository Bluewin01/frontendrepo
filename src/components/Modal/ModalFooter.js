import React from "react";
import { Modal, Button } from "react-bootstrap";
import "./_ModalStyle.css";
export const ModalFooter = (props) => {
  const { buttons } = props;

  return (
    <Modal.Footer>
      {buttons.map((button) => {
        return (
          <Button
            data-testid={button.testId}
            key={button.name}
            onClick={button.handler}
            variant={button.color}
            className="Button"
          >
            {button.name}
          </Button>
        );
      })}
    </Modal.Footer>
  );
};
