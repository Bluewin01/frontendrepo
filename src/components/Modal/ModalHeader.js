import React from "react";
import { Modal } from "react-bootstrap";
import "./_ModalStyle.css";

export const ModalHeader = (props) => {
  const { headerTitle } = props;

  return (
    <Modal.Header className="modalHeader">
      <div className="mx-auto">{headerTitle}</div>
    </Modal.Header>
  );
};
