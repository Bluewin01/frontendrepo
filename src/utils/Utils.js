// test-utils.jsx
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore } from "../store/reducers";

export function render(
  ui,
  { preloadedState, store = setupStore(preloadedState), ...renderOptions } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";
