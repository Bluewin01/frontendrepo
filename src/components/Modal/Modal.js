import React from "react";
import { Modal as ReactModal, Button } from "react-bootstrap";
import "./_ModalStyle.css";
import { ModalHeader } from "./ModalHeader";
import { ModalBody } from "./ModalBody";
import { ModalFooter } from "./ModalFooter";

export const Modal = (props) => {
  const { headerTitle, buttons, children } = props;

  return (
    <>
      <ReactModal
        show={true}
        backdrop="static"
        keyboard={false}
        animation={false}
        dialogClassName={children ? "fullScreenModal" : null}
        centered
      >
        <ModalHeader headerTitle={headerTitle} />
        {children && <ModalBody content={children} />}
        <ModalFooter buttons={buttons} />
      </ReactModal>
    </>
  );
};
