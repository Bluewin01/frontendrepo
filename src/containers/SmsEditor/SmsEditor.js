import "../../assets/css/editor.css";
import React, { useState, useEffect } from "react";
import { Form, Container, Button } from "react-bootstrap";
import { connect } from "react-redux";
import * as TemplateActions from "../../store/Template/TemplateActions";
import PropTypes from "prop-types";
import { Modal } from "../../components/Modal/Modal";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import Loader from "../../components/Loader/Loader";
import History from "../../routes/History";
import { createStructuredSelector } from "reselect";
import {
  makeSelectResponse,
  makeSelectError,
  makeSelectTemplateObject,
} from "../../store/Template/TemplateReselect";
import {
  validateIsInputEmpty,
  validateIsInputListEmpty,
} from "../../utils/InputValidation";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { MultiInputField } from "../../components/MultiInputField/MultiInputField";

const SmsEditor = (props) => {
  const {
    initiateSaveTemplate,
    response,
    error,
    initiateSendSms,
    initiateSetResponseToDefault,
    initiateSetErrorToDefault,
    initiateEditTemplate,
    templateObject,
  } = props;

  const [loading, showLoader] = useState(false);
  const [name, setName] = useState("");
  const [subject, setSubject] = useState("");
  const [description, setDescription] = useState("");
  const [value, setValue] = useState("");
  const [nameError, setNameError] = useState("");
  const [subjectError, setSubjectError] = useState("");
  const [descriptionError, setDescriptionError] = useState("");
  const [valueError, setValueError] = useState("");
  const [showSaveModal, setShowSaveModal] = useState(false);
  const [showCancelModal, setShowCancelModal] = useState(false);
  const [showTestModal, setShowTestModal] = useState(false);
  const [toList, setToList] = useState([{}]);
  const [toCounter, setToCounter] = useState(1);

  const buttons = [
    {
      name: "Test",
      color: "danger",
      testId: "showTestModal",
      handler: () => {
        handleShowTestModal();
      },
    },
    {
      name: "Save",
      color: "danger",
      testId: "showSaveModal",
      handler: () => {
        setShowSaveModal(true);
      },
    },
    {
      name: "Cancel",
      color: "light",
      testId: "showCancelModal",
      handler: () => {
        setShowCancelModal(true);
      },
    },
  ];

  const initiateErrorMessage = () => {
    setNameError("");
    setSubjectError("");
    setDescriptionError("");
    setValueError("");
  };

  const initiateToListError = () => {
    setToList(() => {
      toList.forEach((list) => {
        list.error = "";
      });
      return [...toList];
    });
  };

  useEffect(() => {
    if (templateObject !== null) {
      const { name, subject, description, value } = templateObject;
      setName(name);
      setSubject(subject);
      setDescription(description);
      setValue(value);
    } else {
      setName("");
      setSubject("");
      setDescription("");
      setValue("");
    }
  }, [templateObject]);

  useEffect(() => {
    if (error) toast.error(error);
    if (response) toast.success(response);
    initiateSetErrorToDefault();
    initiateSetResponseToDefault();
  }, [error, response]);

  useEffect(() => {
    setToList([
      {
        error: "",
        name: `To0`,
        value: "",
        placeholder: "e.g: +6010......",
        required: true,
        setInput: setToList,
      },
    ]);
  }, []);

  const handleSave = () => {
    initiateErrorMessage();
    const type = "Sms";
    const validationResult = validateIsInputEmpty({
      name,
      subject,
      description,
      value,
    });
    if (validationResult.length === 0) {
      if (templateObject !== null) {
        initiateEditTemplate(name, subject, description, value, type);
      } else {
        initiateSaveTemplate(name, subject, description, value, type);
      }
      showLoader(true);
      setTimeout(() => {
        showLoader(false);
        History.push("/home");
      }, 1000);
    } else {
      if (validationResult.includes("name"))
        setNameError("Template Name is required.");
      if (validationResult.includes("subject"))
        setSubjectError("Template Subject is required.");
      if (validationResult.includes("description"))
        setDescriptionError("Template Description is required.");
      if (validationResult.includes("value"))
        setDescriptionError("Message is required.");
    }
    setShowSaveModal(false);
  };

  const handleCancel = () => {
    History.push("/home");
  };

  const handleShowTestModal = () => {
    initiateErrorMessage();
    initiateToListError();
    const validationResult = validateIsInputEmpty({ subject, value });
    if (validationResult.length === 0) {
      setShowTestModal(true);
    } else {
      if (validationResult.includes("subject"))
        setSubjectError("Template Subject is required.");
      if (validationResult.includes("value"))
        setValueError("Message is required.");
    }
  };

  const handleSendSms = () => {
    initiateErrorMessage();
    initiateToListError();

    const emptyListValidation = validateIsInputListEmpty(toList);

    if (emptyListValidation.length === 0) {
      let phoneList = [];
      toList.forEach((list) => {
        phoneList.push(list.value);
      });
      initiateSendSms(phoneList, value, subject);
    } else {
      if (emptyListValidation.length > 0) {
        setToList(() => {
          emptyListValidation.forEach((fieldName) => {
            let foundIndex = toList.findIndex(
              (list) => list.name === fieldName
            );
            toList[foundIndex].error = "Phone number is required";
          });
          return [...toList];
        });
      }
    }
  };

  const handleCreateToField = () => {
    setToList([
      ...toList,
      {
        error: "",
        name: `To${toCounter}`,
        value: "",
        placeholder: "eg: +6010.....",
        setInput: setToList,
      },
    ]);
    setToCounter(toCounter + 1);
  };

  const handleDeleteToField = (fieldName) => {
    setToList(toList.filter((field) => field.name !== fieldName));
  };

  const toField = {
    label: "To",
    inputFields: toList,
    required: true,
    createField: handleCreateToField,
    deleteField: handleDeleteToField,
  };

  return (
    <Container>
      <ToastContainer
        position="top-right"
        autoClose={5000}
        hideProgressBar
        newestOnTop={false}
        closeOnClick
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover={false}
      />
      {loading ? <Loader></Loader> : null}
      <Form>
        <br></br>
        <h4 className="text-center mb-5">SMS Template Editor</h4>
        <div className="row">
          <Form.Group className="col mb-6" controlId="template_name">
            <Form.Label className="required">Template Name</Form.Label>
            <Form.Control
              type="text"
              name="name"
              data-testid="name"
              value={name}
              onChange={(e) => {
                setName(e.target.value);
                if (validateIsInputEmpty(e.target.value).length === 0)
                  setNameError("");
              }}
              autoComplete="off"
            />
            <ErrorMessage error={nameError} />
          </Form.Group>
          <Form.Group className="col mb-6" controlId="message_sub">
            <Form.Label className="required">Template Subject</Form.Label>
            <Form.Control
              type="text"
              name="subject"
              data-testid="subject"
              value={subject}
              onChange={(e) => {
                setSubject(e.target.value);
                if (validateIsInputEmpty(e.target.value).length === 0)
                  setSubjectError("");
              }}
              autoComplete="off"
            />

            <ErrorMessage error={subjectError} />
          </Form.Group>
        </div>
        <Form.Group className="mb-3" controlId="description">
          <Form.Label className="required">Template Description</Form.Label>
          <Form.Control
            as="textarea"
            maxLength="30"
            rows={3}
            data-testid="description"
            value={description}
            onChange={(e) => {
              setDescription(e.target.value);
              if (validateIsInputEmpty(e.target.value).length === 0)
                setDescriptionError("");
            }}
            autoComplete="off"
          />
          <ErrorMessage error={descriptionError} />
        </Form.Group>
        <Form.Group className="mb-3" controlId="message">
          <Form.Label className="required">Message</Form.Label>
          <Form.Control
            as="textarea"
            maxLength="160"
            placeholder="Max Character 160"
            rows={3}
            data-testid="message"
            value={value}
            onChange={(e) => {
              setValue(e.target.value);
              if (validateIsInputEmpty(e.target.value).length === 0)
                setValueError("");
            }}
            autoComplete="off"
          />
          <ErrorMessage error={valueError} />
        </Form.Group>
        <div className="text-right mt-3">
          {buttons.map((button) => {
            return (
              <Button
                data-testid={button.testId}
                key={button.name}
                className="Button"
                variant={button.color}
                onClick={button.handler}
              >
                {button.name}
              </Button>
            );
          })}
        </div>
      </Form>
      {showTestModal && (
        <Modal
          headerTitle={`Test SMS Template`}
          buttons={[
            {
              name: "Send SMS",
              color: "danger",
              handler: handleSendSms,
            },
            {
              name: "Back",
              color: "light",
              handler: () => setShowTestModal(false),
            },
          ]}
        >
          <MultiInputField {...toField} />
        </Modal>
      )}
      {showSaveModal && (
        <Modal
          headerTitle={`Are you sure you want to save SMS template`}
          buttons={[
            {
              name: "Confirm",
              color: "danger",
              testId: "triggerHandler",
              handler: handleSave,
            },
            {
              name: "Cancel",
              color: "light",
              testId: "closeModal",
              handler: () => setShowSaveModal(false),
            },
          ]}
        />
      )}
      {showCancelModal && (
        <Modal
          headerTitle={`All changes will be lost. Are you sure to continue?`}
          buttons={[
            {
              name: "Confirm",
              color: "danger",
              handler: handleCancel,
            },
            {
              name: "Cancel",
              color: "light",
              handler: () => setShowCancelModal(false),
            },
          ]}
        />
      )}
    </Container>
  );
};

const mapStateToProps = createStructuredSelector({
  error: makeSelectError(),
  response: makeSelectResponse(),
  templateObject: makeSelectTemplateObject(),
});

const mapDispatchToProps = (dispatch) => {
  return {
    initiateSaveTemplate: (name, subject, description, value, type) =>
      dispatch(
        TemplateActions.saveTemplateRequest({
          name,
          subject,
          description,
          value,
          type,
        })
      ),
    initiateSendSms: (phoneList, value, subject) =>
      dispatch(
        TemplateActions.sendSmsRequest({
          phoneList,
          value,
          subject,
        })
      ),
    initiateSetResponseToDefault: () =>
      dispatch(TemplateActions.setResponseToDefault()),
    initiateSetErrorToDefault: () =>
      dispatch(TemplateActions.setErrorToDefault()),
    initiateEditTemplate: (name, subject, description, value, type) =>
      dispatch(
        TemplateActions.editTemplateRequest({
          name,
          subject,
          description,
          value,
          type,
        })
      ),
  };
};

SmsEditor.propTypes = {
  initiateSaveTemplate: PropTypes.func,
  error: PropTypes.string,
  response: PropTypes.string,
  initiateSendSms: PropTypes.func,
  initiateSetResponseToDefault: PropTypes.func,
  initiateSetErrorToDefault: PropTypes.func,
  initiateEditTemplate: PropTypes.func,
  templateObject: PropTypes.object,
};

export default connect(mapStateToProps, mapDispatchToProps)(SmsEditor);
