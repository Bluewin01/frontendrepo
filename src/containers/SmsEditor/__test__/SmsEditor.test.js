import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { screen, cleanup, fireEvent, act } from "@testing-library/react";
import { render } from "../../../utils/Utils";
import renderer from "react-test-renderer";
import store from "../../../store";
import { Provider } from "react-redux";
import SmsEditor from "../SmsEditor";
import { Modal } from "../../../components/Modal/Modal";

jest.useFakeTimers();

describe("Containers: SMS Editor", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  afterEach(cleanup);

  describe("Positive Case", () => {
    it("should be render correctly", () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <SmsEditor />
          </Provider>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should be able to render email page with error", () => {
      render(<SmsEditor />, {
        preloadedState: {
          Template: {
            error: "fakeError",
            response: null,
            templateObject: null,
          },
        },
      });
    });

    it("should be able to render email page with response", () => {
      render(<SmsEditor />, {
        preloadedState: {
          Template: {
            error: null,
            response: "fakeResponse",
            templateObject: null,
          },
        },
      });
    });

    it("should can input all the fields", () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      expect(inputName.value).toBe(mockTemplateName);
      expect(inputSubject.value).toBe(mockSubject);
      expect(inputDescription.value).toBe(mockDescription);
      expect(inputMessage.value).toBe(mockMessage);
    });

    it("should can be save the template after input all fields", async () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);
      act(() => jest.runOnlyPendingTimers());

      expect(saveButton).toHaveTextContent("Save");
      expect(confirmButton).toHaveTextContent("Confirm");
    });

    it("should can be test send the template after input all fields", async () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";
      const mockRecipient = "Mock Recipient";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      const sendButton = screen.getAllByRole("button")[0];
      fireEvent.click(sendButton);

      const recipient = screen.getAllByRole("textbox")[0];
      const subject = screen.getAllByRole("textbox")[1];

      fireEvent.change(recipient, {
        target: { value: mockRecipient },
      });

      const confirmButton = screen.getAllByRole("button")[0];
      fireEvent.click(confirmButton);

      expect(sendButton).toHaveTextContent("Test");
      expect(confirmButton).toHaveTextContent("Send SMS");
    });

    it("should can be edit the template", async () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";
      const mockRecipient = "Mock Recipient";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      const sendButton = screen.getAllByRole("button")[0];
      fireEvent.click(sendButton);

      const recipient = screen.getAllByRole("textbox")[0];
      const subject = screen.getAllByRole("textbox")[1];

      fireEvent.change(recipient, {
        target: { value: mockRecipient },
      });

      const confirmButton = screen.getAllByRole("button")[0];
      fireEvent.click(confirmButton);

      expect(sendButton).toHaveTextContent("Test");
      expect(confirmButton).toHaveTextContent("Send SMS");
    });

    it("should be redirect to /home route when click cancel button", () => {
      render(<SmsEditor />);
      const cancelButton = screen.getByText("Cancel");
      fireEvent.click(cancelButton);

      expect(
        screen.getByText("All changes will be lost. Are you sure to continue?")
      ).toBeInTheDocument();

      const confirmButton = screen.getByText("Confirm");
      fireEvent.click(confirmButton);
      expect(global.window.location.href).toContain("/home");
    });

    it("should be shown the cancel modal pop-up", () => {
      render(<SmsEditor />);
      const cancelButton = screen.getByText("Cancel");
      fireEvent.click(cancelButton);

      expect(
        screen.getByText("All changes will be lost. Are you sure to continue?")
      ).toBeInTheDocument();
    });

    it("should be able to save with template object and redirect to home page", async () => {
      render(<SmsEditor />, {
        preloadedState: {
          Template: {
            error: null,
            response: null,
            templateObject: {
              name: "fakeName",
              subject: "fakeSubject",
              description: "fakeDescription",
              value: "fakeMessage",
            },
          },
        },
      });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);

      expect(global.window.location.href).toContain("/home");
    });
  });

  describe("Negative Case", () => {
    it("should show error message for template name when saving with no template name", () => {
      render(<SmsEditor />);
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";
      const mockErrorMessage = "Template Name is required.";

      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);

      const errorMessage = screen.getByTestId("errorMessage");
      expect(errorMessage).toHaveTextContent(mockErrorMessage);
    });

    it("should show error message for template subject when saving with no template subject", () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockDescription = "Mock Template Description";
      const mockMessage = "Mock Template Message";
      const mockErrorMessage = "Template Subject is required.";

      const inputName = screen.getByTestId("name");
      const inputDescription = screen.getByTestId("description");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });
      fireEvent.change(inputMessage, { target: { value: mockMessage } });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);

      const errorMessage = screen.getByTestId("errorMessage");
      expect(errorMessage).toHaveTextContent(mockErrorMessage);
    });

    it("should show error message for template subject when saving with no template message", () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockDescription = "Mock Template Description";
      const mockErrorMessage = "Message is required.";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputDescription = screen.getByTestId("description");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputDescription, {
        target: { value: mockDescription },
      });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);

      const errorMessage = screen.getByTestId("errorMessage");
      expect(errorMessage).toHaveTextContent(mockErrorMessage);
    });
    it("should show error message for template subject when saving with no template description", () => {
      render(<SmsEditor />);
      const mockTemplateName = "Mock Template Name";
      const mockSubject = "Mock Template Subject";
      const mockMessage = "Mock Template Message";
      const mockErrorMessage = "Template Description is required.";

      const inputName = screen.getByTestId("name");
      const inputSubject = screen.getByTestId("subject");
      const inputMessage = screen.getByTestId("message");

      fireEvent.change(inputName, { target: { value: mockTemplateName } });
      fireEvent.change(inputSubject, { target: { value: mockSubject } });
      fireEvent.change(inputMessage, {
        target: { value: mockMessage },
      });

      const saveButton = screen.getByTestId("showSaveModal");
      fireEvent.click(saveButton);

      const confirmButton = screen.getByTestId("triggerHandler");
      fireEvent.click(confirmButton);

      const errorMessage = screen.getByTestId("errorMessage");
      expect(errorMessage).toHaveTextContent(mockErrorMessage);
    });
    it("should go to homepage when click cancel", () => {
      render(<SmsEditor />);
      const mockButton = screen.getAllByRole("button")[2];
      expect(mockButton).toBeDefined();
      expect(mockButton).toHaveTextContent("Cancel");
      fireEvent.click(mockButton);
      render(
        <Modal
          headerTitle={`All changes will be lost. Are you sure to continue?`}
          showBody={false}
          buttons={[
            {
              name: "Confirm",
              color: "danger",
            },
            {
              name: "Cancel",
              color: "light",
            },
          ]}
        />
      );
      const confirmButton = screen.findAllByText("Confirm");
      expect(confirmButton).toBeDefined();
    });
  });
});
