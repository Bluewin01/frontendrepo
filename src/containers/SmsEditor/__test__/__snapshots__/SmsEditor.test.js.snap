// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers: SMS Editor Positive Case should be render correctly 1`] = `
<div
  className="container"
>
  <div
    className="Toastify"
  />
  <form
    className=""
  >
    <br />
    <h4
      className="text-center mb-5"
    >
      SMS Template Editor
    </h4>
    <div
      className="row"
    >
      <div
        className="col mb-6 form-group"
      >
        <label
          className="required form-label"
          htmlFor="template_name"
        >
          Template Name
        </label>
        <input
          autoComplete="off"
          className="form-control"
          data-testid="name"
          id="template_name"
          name="name"
          onChange={[Function]}
          type="text"
          value=""
        />
        
      </div>
      <div
        className="col mb-6 form-group"
      >
        <label
          className="required form-label"
          htmlFor="message_sub"
        >
          Template Subject
        </label>
        <input
          autoComplete="off"
          className="form-control"
          data-testid="subject"
          id="message_sub"
          name="subject"
          onChange={[Function]}
          type="text"
          value=""
        />
        
      </div>
    </div>
    <div
      className="mb-3 form-group"
    >
      <label
        className="required form-label"
        htmlFor="description"
      >
        Template Description
      </label>
      <textarea
        autoComplete="off"
        className="form-control"
        data-testid="description"
        id="description"
        maxLength="30"
        onChange={[Function]}
        rows={3}
        value=""
      />
      
    </div>
    <div
      className="mb-3 form-group"
    >
      <label
        className="required form-label"
        htmlFor="message"
      >
        Message
      </label>
      <textarea
        autoComplete="off"
        className="form-control"
        data-testid="message"
        id="message"
        maxLength="160"
        onChange={[Function]}
        placeholder="Max Character 160"
        rows={3}
        value=""
      />
      
    </div>
    <div
      className="text-right mt-3"
    >
      <button
        className="Button btn btn-danger"
        data-testid="showTestModal"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Test
      </button>
      <button
        className="Button btn btn-danger"
        data-testid="showSaveModal"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
      <button
        className="Button btn btn-light"
        data-testid="showCancelModal"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
